### PR TITLE
[ASDisplayNode] Relinquish Lock Before Calling setNeedsLayout When Applying Pending State

### PR DESCRIPTION
**Pull request by Adlai-Holler**
Alternative to #2147 . We need to hold the lock the entire time that we're accessing the pending state.

### CODE REVIEW

File: AsyncDisplayKit/ASDisplayNode.mm [hunk 1]
```diff
@@ -1176,23 +1176,28 @@ - (void)setContentsScaleForDisplay:(CGFloat)contentsScaleForDisplay
...
 
+    [_pendingViewState clearChanges];
+  }
+
+  // NOTE: It's important to call this without holding the lock.
   // FIXME: Ideally we'd call this as soon as the node receives -setNeedsLayout
   // but automatic subnode management would require us to modify the node tree
   // in the background on a loaded node, which isn't currently supported.
-  if (_pendingViewState.hasSetNeedsLayout) {
+  if (shouldSetNeedsLayout) {
```
Comment (garrettmoon) [inline AsyncDisplayKit/ASDisplayNode.mm]: Is it ok that setNeedsLayout has moved to after the pending state has been applied? Currently setNeedLayout sometimes lays out synchronously?

Comment (Adlai-Holler) [inline AsyncDisplayKit/ASDisplayNode.mm]: Yeah it's OK – in fact it's probably better to have all the view/layer attributes updates before the layout pass. Any thoughts about risk here?

Comment (garrettmoon) [inline AsyncDisplayKit/ASDisplayNode.mm]: I think the biggest risk is moving the call ordering, otherwise, LGTM!

Comment (Adlai-Holler) [inline AsyncDisplayKit/ASDisplayNode.mm]: You're right, we should call __setNeedsLayout before potentially calling setNeedsLayout on the view/layer to be safe.

**Pull request closed without merging**